src/economy/economy.cc

```diff
@@ -526,6 +526,64 @@ bool Economy::needs_ware_or_worker(DescriptionIndex const ware_or_worker_type) c
 	}
 }
 
+bool Economy::has_building(const DescriptionIndex di) const {
+	const std::string& name = owner().tribe().get_building_descr(di)->name();
+	for (const auto& bld : owner().get_building_statistics(di)) {
+		if (!bld.is_constructionsite) {
+			if (BaseImmovable* i = owner().egbase().map()[bld.pos].get_immovable()) {
+				if (i->descr().name() == name) {
+					if (dynamic_cast<const Building&>(*i).get_economy(type_) == this) {
+						return true;
```

Comment:
How often do we have to iterate over these buildings in average?

Reply:
This function is currently used only when a player clicks on the flag, to determine whether the flag's economy has a scout's house or not. Every `Player` object keeps stats on owned buildings, and players usually have just one big economy (or two, in shared kingdom mode) and very few scout's houses. So for all practical situations this is called not often and is efficient.